### PR TITLE
SocketGatheringWriteTest improvements

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
SocketGatherWriteTest has been observed to fail and it has numerous issues which when resolved may help reduce the test failures.

Modifications:
- A volatile counter and a spin/sleep loop is used to trigger test termination. Incrementing a volatile is generally bad practice and can be avoided in this situation. This mechanism can be replaced by a promise. This mechanism should also trigger upon exception or channel inactive.
- The TestHandler maintains an internal buffer, but it is not released. We now only create a buffer on the server side and release it after comparing the expected results.
- The composite buffer creation logic can be simplified, also the existing composite buffer doesn't take into account the buffer's reader index when building buf2.

Result:
Cleaner test.

### CODE REVIEW

**Comment by Scottmitch** [comment]
Discovered this while investigating https://github.com/netty/netty/issues/7062

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java [hunk 1]
```diff
@@ -175,49 +166,84 @@ private void testGatheringWrite0(
...
-                Throwable cause) throws Exception {
+        public final void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
             if (exception.compareAndSet(null, cause)) {
-                ctx.close();
+                handleException(ctx, cause);
             }
+            super.exceptionCaught(ctx, cause);
+        }
+
+        void handleException(ChannelHandlerContext ctx, Throwable cause) {
```
Comment (netkins) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/SocketGatheringWriteTest.java]: MAJOR Remove this unused method parameter "cause". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by Scottmitch** [comment]
Build failure is attributed to https://github.com/netty/netty/issues/7056

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (0bf614d9e9395320f1712fc8e47b245df648fee0) and 4.0 (a044f1705a)

**Pull request closed without merging**